### PR TITLE
Implementation of the Pivot table component

### PR DESCRIPTION
**Pull request by RiadhCherni**
Add pivot table component to Galite

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/MPivotTable.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2013-2022 kopiLeft Services SARL, Tunis TN
+ * Copyright (c) 1990-2022 kopiRight Managed Solutions GmbH, Wien AT
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/MPivotTable.kt:3]: Replace `2022` with `2023`

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/MPivotTable.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  // --------------------------------------------------------------------
+  // DATA MEMBERS
+  // --------------------------------------------------------------------
+  // Columns contains all columns defined by the user
+  var columns = mutableListOf<VPivotTableColumn?>()    // array of column definitions
+
+
+  // Baserows contains data give by the request of the user
+  internal var userRows: ArrayList<VPivotTableRow>? = ArrayList(500)
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/MPivotTable.kt:35]: Why is a limit of 500? What is the expected treatment when we have more than 500 data rows?

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VHelpGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2013-2022 kopiLeft Services SARL, Tunis TN
+ * Copyright (c) 1990-2022 kopiRight Managed Solutions GmbH, Wien AT
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VHelpGenerator.kt:3]: Replace `2022` with `2023`

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VHelpGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  // ----------------------------------------------------------------------
+  // ACCESSORS
+  // ----------------------------------------------------------------------
+  /**
+   * prints a compilation unit
+   */
+  fun helpOnPivotTable(name: String,
+                      commands: List<VCommand>,
+                      model: MPivotTable,
+                      help: String?): String? {
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VHelpGenerator.kt:45]: Fix indentation

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+package org.kopi.galite.visual.pivotTable
+
+import org.kopi.galite.util.base.InconsistencyException
+import org.kopi.galite.visual.*
+import org.kopi.galite.visual.form.VConstants
+import org.kopi.galite.visual.l10n.LocalizationManager
+import org.kopi.galite.visual.dsl.common.Trigger
+import java.io.File
+import java.net.MalformedURLException
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt:27]: Order imports

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    }
+  }
+
+  // ----------------------------------------------------------------------
+  // DATA MEMBERS
+  // ----------------------------------------------------------------------
+  val model: MPivotTable = MPivotTable()
+  private var built = false
+  private var pageTitle = ""
+  var pivottableType = Constants.DEFAULT_RENDERER  // Default pivot table type
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt:54]: rename `pivotTableType` with `defaultRenderer`

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+  override fun getType() = org.kopi.galite.visual.Constants.MDL_PIVOT_TABLE
+
+  /**
+   * Close window
+   */
+  @Deprecated("call method in display; model must not be closed")
+  fun close() {
+    getDisplay()!!.closeWindow()
+  }
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt:70]: why do we add a deprecated method to a new module?

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+  override fun destroyModel() {
+    super.destroyModel()
+  }
+
+  /**
+   * Sets the new type of this pivot table model.
+   * @param type The new pivot table type.
+   */
+  internal fun setType(type: String) {
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/pivotTable/VPivotTable.kt:80]: Do we mean renderer by type? if so rename to `setDefaultRenderer`.

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+class DPivotTable(private val pivotTable: VPivotTable) : DWindow(pivotTable), UPivotTable {
+
+  //---------------------------------------------------
+  // DATA MEMBERS
+  //---------------------------------------------------
+  private val model: MPivotTable = pivotTable.model // pivot table model
+  private val pivotData = PivotTable.PivotData()
+  private val pivotOptions = PivotTable.PivotOptions()
+  private val listeDimensionRows = mutableListOf<String>()
+  private val listeDimensionColumns = mutableListOf<String>()
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt:39]: Rename `listeDimensionRows` to `rows`
Rename `listeDimensionColumns` to `columns`

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  // IMPLEMENTATIONS
+  //---------------------------------------------------
+  override fun run() {
+    pivotTable.initPivotTable()
+    pivotTable.setMenu()
+  }
+
+  override fun build() {
+    model.columns
+      .forEach {
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt:55]: Remove new line & fix following indentation

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+          listeDimensionRows.add(it.label)
+        }
+        if (it?.position == Position.DIMENSION_COL) {
+          listeDimensionColumns.add(it.label)
+        }
+      }
+
+    model.userRows
+      ?.flatMap { it.data.filterNotNull() }
+      ?.chunked(model.columns.count()) { rows ->
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt:67]: Remove new lines & fix following indentation

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    pivotOptions.setAggregator(pivotTable.aggregator.first, pivotTable.aggregator.second)
+
+    // Pivot table renderer
+    if(listeDimensionRows.isNotEmpty() && listeDimensionColumns.isNotEmpty()) {
+      pivotOptions.setCharts(true)
+    }
+    pivotOptions.setDisabledRenderers(*pivotTable.disabledRerenders.toTypedArray())
+
+    // Pivot table mode
+    val pivotMode = if(pivotTable.interactive == 0) {
```
Comment (mgrati) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/pivotTable/DPivotTable.kt:88]: add space after if

File: galite-demo/galite-vaadin/src/main/kotlin/org/kopi/galite/demo/bill/BillForm.kt
Comment (mgrati) [inline galite-demo/galite-vaadin/src/main/kotlin/org/kopi/galite/demo/bill/BillForm.kt:1]: In galite-demo, I suggest focusing on creating one pivot table that can show the maximum allowed functionalities developed in this module.

I think we can use the Client form to do this, with a report that gets data from tables `CLIENTS` and `PURCHASE`.

The pivot table can show us the purchases per client/ country / city ...
I also suggest for a better visibilty to add more data to these tables using org.kopi.galite.demo.database.Migration

**Review by mgrati — CHANGES_REQUESTED** [review]
I added few code style comments below. Please check.

The task `./gradlew test` is failed. please check and fix the failed tests. The failed test is in Class org.kopi.galite.tests.ui.vaadin.form.FormTests and it is caused by the changes in Migration.addProducts. You can fix it by extracting the first 4 records of the loop and start your loop from 4 

 I also think that some changes need to be done for Galite demo : 
- The products pivot table is OK, However, the pivot tables added for clients or bill products I don't find particularly meaningful.
- For the client pivot table, I suggest that you add a left join with the table purchase and maybe product so you can add quantity and price as measures. and at the same time check if dimensions and measures can accept null values.
- I think we can delete the bill product pivot table since it's not really meaningful

File: galite-core/src/main/kotlin/org/kopi/galite/visual/dsl/pivottable/PivotTable.kt [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+  fun <T> trigger(vararg pivotTriggerEvents: PivotTableTriggerEvent<T>, method: () -> T): Trigger {
+    val event = formEventList(pivotTriggerEvents)
+    val pivotTableAction = Action(null, method)
+    val trigger = FormTrigger(event, pivotTableAction)
+
+    triggers.add(FormTrigger(event, Action(null, method)))
+
+    // PIVOT TABLE TRIGGERS
+    triggers.forEach { trigger ->
+
```
[the remark this answers is not in the document]
Comment (RiadhCherni) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/dsl/pivottable/PivotTable.kt:102]: done

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/Constants.kt [hunk 14]
```diff
@@ -20,52 +20,61 @@ package org.kopi.galite.visual.pivottable
 
 import org.kopi.galite.visual.form.VConstants
 import org.kopi.galite.visual.Constants
+import org.vaadin.addons.componentfactory.PivotTable.Renderer
+import org.vaadin.addons.componentfactory.PivotTable.Aggregator
```
[the remark this answers is not in the document]
Comment (RiadhCherni) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/Constants.kt:24]: done

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+abstract class VPivotTable internal constructor() : VWindow(), VConstants {
+  companion object {
+
+    init {
+      WindowController.windowController.registerWindowBuilder(
+        org.kopi.galite.visual.Constants.MDL_PIVOT_TABLE,
+        object : WindowBuilder {
+          override fun createWindow(model: VWindow): UWindow {
+            return UIFactory.uiFactory.createView(
+              model) as UPivotTable
```
[the remark this answers is not in the document]
Comment (RiadhCherni) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt:43]: done

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+
+import java.io.File
+import java.net.MalformedURLException
+
+import org.kopi.galite.util.base.InconsistencyException
+import org.kopi.galite.visual.*
+import org.kopi.galite.visual.form.VConstants
+import org.kopi.galite.visual.l10n.LocalizationManager
+import org.kopi.galite.visual.dsl.common.Trigger
+import org.vaadin.addons.componentfactory.PivotTable.*
```
[the remark this answers is not in the document]
Comment (RiadhCherni) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt:29]: Dans l'ensemble du projet, les imports de galite précédent toujours les imports de vaadin.

[the remark this answers is not in the document]
Comment (mgrati) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt:29]: That does not mean it's the right way. Please check our style guide references : http://mediawiki.dms.at/wiki/KOPI_Style_Guide#Package_and_Import_Statements 
image

File: galite-core/src/main/kotlin/org/kopi/galite/visual/dsl/pivottable/Dimension.kt [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  // ----------------------------------------------------------------------
+  // POSITION CLASS
+  // ----------------------------------------------------------------------
+
+  enum class Position(val value: Int) {
+      NONE(Constants.DIMENSION_NO_POSITION),
+      ROW(Constants.DIMENSION_ROW),
+      COLUMN(Constants.DIMENSION_COLUMN),
```
[the remark this answers is not in the document]
Comment (RiadhCherni) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/dsl/pivottable/Dimension.kt]: done

File: galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (RiadhCherni) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/pivottable/VPivotTable.kt]: done

**Review by mgrati — APPROVED** [review]

**Pull request merged**